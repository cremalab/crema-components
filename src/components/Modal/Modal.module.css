.modalOverlay {
  background-color: hsla(170, 5%, 15%, 0.8);
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.modalWindow {
  width: 375px;
  border-radius: 5px;
  background-color: #fff;
  margin: 100px auto;
  padding: 8px 20px 8px 20px;
}

.modalTitle {
  display: flex;
  justify-content: space-between;
}

.modalTitleClose {
  color: #808080;
  cursor: pointer;
}

@media screen and (max-width: 500px) {
  .modalWindow {
    width: 250px;
    height: 250px;
  }
}
