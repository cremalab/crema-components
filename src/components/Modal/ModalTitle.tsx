import { ReactNode } from "react"
import styles from "./Modal.module.css"

interface Props {
  children: ReactNode
  onClose?: () => unknown
}

export function ModalTitle({ children, onClose }: Props) {
  return (
    <div className={styles.modalTitle}>
      <div>{typeof children === "string" ? <p>{children}</p> : children}</div>
      {onClose ? (
        <div>
          <div
            className={styles.modalTitleClose}
            aria-label="close the modal"
            onClick={onClose}
            onKeyPress={onClose}
            role="button"
            tabIndex={0}
          >
            x
          </div>
        </div>
      ) : null}
    </div>
  )
}
