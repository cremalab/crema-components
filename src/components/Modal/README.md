# `<Modal />`

DESCRIPTION_HERE

## Directory Structure

- `Modal.stories.tsx`: Component stories (`npm run test:playground`)
- `Modal.test.tsx`: Component tests (`npm run test:unit`)
- `Modal.tsx`: Component code
- `index.ts`: Component export
- `README.md`: Component documentation (hey, that's me!)
