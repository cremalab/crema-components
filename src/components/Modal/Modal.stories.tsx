import { ComponentMeta, Story } from "@storybook/react"
import { ComponentProps, useState } from "react"
import { Button } from "../Button"
import { TextInput } from "../TextInput"
import { Modal } from "./Modal"

/**
 * See Storybook Docs: Writing Stories
 * https://storybook.js.org/docs/basics/writing-stories/
 */

export default {
  title: "Components/Modal",
  component: Modal,
} as ComponentMeta<typeof Modal>

export const Basic: Story<ComponentProps<typeof Modal>> = (args) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setIsOpen(true)}>
        Open basic modal
      </Button>
      <Modal open={isOpen} onClose={() => setIsOpen(false)} {...args}>
        <p>I am a basic modal ðŸ˜ƒ</p>
        <Button
          ariaLabel="Close Modal"
          aria-labelledby="close-modal"
          className="close-modal"
          onClick={() => setIsOpen(false)}
        >
          Close
        </Button>
      </Modal>
    </div>
  )
}

export const WithTitle: Story<ComponentProps<typeof Modal>> = (args) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setIsOpen(true)}>
        Open modal with title
      </Button>
      <Modal
        title="Modal With Title"
        open={isOpen}
        onClose={() => setIsOpen(false)}
        {...args}
      >
        <p>I am a modal. I have a title! ðŸ˜ƒ</p>
        <Button
          ariaLabel="Close Modal"
          aria-labelledby="close-modal"
          className="close-modal"
          onClick={() => setIsOpen(false)}
        >
          Close
        </Button>
      </Modal>
    </div>
  )
}

export const WithMoreContent: Story<ComponentProps<typeof Modal>> = (args) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setIsOpen(true)}>
        Open modal with more content
      </Button>
      <Modal
        title="Modal With Title"
        open={isOpen}
        onClose={() => setIsOpen(false)}
        {...args}
      >
        <div>
          <p>Modals can be cool</p>
          <p>I am a modal. I have a title! ðŸ˜ƒ</p>
          <TextInput name="test-input" label="Test Input" hideLabel={false} />
          <div style={{ marginTop: "1rem" }}>
            <Button
              ariaLabel="Close Modal"
              aria-labelledby="close-modal"
              className="close-modal"
              onClick={() => setIsOpen(false)}
            >
              Close
            </Button>
          </div>
        </div>
      </Modal>
    </div>
  )
}
