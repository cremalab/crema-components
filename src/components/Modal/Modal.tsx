import { HTMLProps, ReactNode, useEffect, useRef } from "react"
import { createPortal } from "react-dom"
import FocusLock from "react-focus-lock"
import { useKeyPress } from "../../hooks/useKeyPress"
import { useOutsideClick } from "../../hooks/useOutsideClick"
import styles from "./Modal.module.css"
import { ModalTitle } from "./ModalTitle"

interface Props extends HTMLProps<HTMLDivElement> {
  children: ReactNode
  onClose?: () => void
}

export function Modal({ children, open, onClose, title, ...props }: Props) {
  const modalRef = useRef<HTMLDivElement>(null)

  useOutsideClick(modalRef, () => {
    onClose?.()
  })

  useKeyPress(["Escape"], () => {
    if (open) onClose?.()
  })

  useEffect(() => {
    if (open) {
      document.body.style.overflow = "hidden"
    } else {
      document.body.style.overflow = "auto"
    }
  }, [open])

  if (!open) return null

  return createPortal(
    <div className={styles.modalOverlay} data-testid="modal-overlay">
      <div
        className={styles.modalWindow}
        ref={modalRef}
        aria-modal
        aria-labelledby={title}
        role="dialog"
        {...props}
      >
        <FocusLock disabled={!open}>
          {title ? <ModalTitle onClose={onClose}>{title}</ModalTitle> : null}
          <div>{children}</div>
        </FocusLock>
      </div>
    </div>,
    document.body,
  )
}
