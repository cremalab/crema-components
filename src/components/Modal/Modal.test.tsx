import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Modal } from "./Modal"
import { ModalTitle } from "./ModalTitle"

describe("Modal", () => {
  it("renders nothing when open = false", async () => {
    // Arrange
    // Act
    const { asFragment } = render(
      <Modal open={false}>this should not render</Modal>,
    )

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders content when open = true", async () => {
    // Arrange
    const title = "title"
    const children = "children"

    // Act
    render(
      <Modal open={true} title={title}>
        {children}
      </Modal>,
    )
    const titleNode = await screen.findByText(title)
    const childrenNode = await screen.findByText(children)

    // Assert
    expect(titleNode).toBeInTheDocument()
    expect(childrenNode).toBeInTheDocument()
  })

  it("calls onClose when the modal overlay is clicked", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Modal open={true} title={title} onClose={onClose}>
        {children}
      </Modal>,
    )
    const overlay = await screen.findByTestId("modal-overlay")
    await userEvent.click(overlay)

    expect(onClose).toHaveBeenCalled()
  })

  it("calls onClose when the escape key is pressed", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Modal open={true} title={title} onClose={onClose}>
        {children}
      </Modal>,
    )

    await userEvent.keyboard("[Escape]")

    expect(onClose).toHaveBeenCalled()
  })

  it("does not call onClose when the modal is not open and the escape key is pressed", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Modal open={false} title={title} onClose={onClose}>
        {children}
      </Modal>,
    )
    await userEvent.keyboard("[Escape]")

    expect(onClose).not.toHaveBeenCalled()
  })

  it("does not render a title if no title is provided", async () => {
    // Arrange
    const children = "children"

    // Act
    render(<Modal open={true}>{children}</Modal>)

    // Assert
    expect(screen.getByText(children)).toBeInTheDocument()
  })
})

describe("ModalTitle", () => {
  it("renders the children", async () => {
    // Arrange
    // Act
    render(<ModalTitle>Test Title</ModalTitle>)
    const received = await screen.findByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
  })

  it("does not render the close button if no onClose is given", async () => {
    // Arrange
    // Act
    render(<ModalTitle>Test Title</ModalTitle>)
    const closeButton = screen.queryByLabelText("close the modal")

    // Assert
    expect(closeButton).toBeNull()
  })

  it("calls onClose when the close button is clicked", async () => {
    // Arrange
    const onClose = jest.fn()

    // Act
    render(<ModalTitle onClose={onClose}>Test Title</ModalTitle>)

    const closeButton = screen.queryByLabelText(
      "close the modal",
    ) as HTMLElement
    await userEvent.click(closeButton)

    // Assert
    expect(onClose).toHaveBeenCalledTimes(1)
  })

  it("renders children in a paragraph tag if children is a string", async () => {
    // Arrange
    const children = "Test Title"
    // Act
    render(<ModalTitle>{children}</ModalTitle>)
    const received = await screen.findByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
    expect(received.tagName).toBe("P")
  })

  it("renders children in a div tag if children is not a string", async () => {
    // Arrange
    const children = <div>Test Title</div>
    // Act
    render(<ModalTitle>{children}</ModalTitle>)
    const received = await screen.findByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
    expect(received.tagName).toBe("DIV")
  })
})
