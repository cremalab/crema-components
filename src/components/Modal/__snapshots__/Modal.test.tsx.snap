// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders nothing when open = false 1`] = `<DocumentFragment />`;
